Run tests in context managers

Sheila-nk opened the pull request:
- plugged in the `DebugDTRunner`
- overrode its `run` function to ensure tests are run in context managers

Closes #82

Sheila-nk commented:
**Bug**
It seems the `DebugDTRunner` implementation is not raising the defined exceptions as it should. This is likely because of this part of the code in the [DoctestRunner run function](https://github.com/python/cpython/blob/23a6db98f21cba3af69a921f01613bd5f602bf6d/Lib/doctest.py#L1472):
```python
       if out is None:
            encoding = save_stdout.encoding
            if encoding is None or encoding.lower() == 'utf-8':
                out = save_stdout.write
            else:
                # Use backslashreplace error handling on write
                def out(s):
                    s = str(s.encode(encoding, 'backslashreplace'), encoding)
                    save_stdout.write(s)
```

From the code implementation here in the [DTRunner](https://github.com/ev-br/scpdt/blob/1eab6a45134dc9c2520ffd7d0cc4ffb89c826cec/scpdt/impl.py#L329)
```python
    def _report_item_name(self, out, item_name, new_line=False):
        if item_name is not None:
            out("\n " + item_name + "\n " + "-"*len(item_name))
            if new_line:
                out("\n")
```
`out` is used as a callable but the raised TypeError indicates that it is still a `list` object

ev-br commented:
What does DebugDTRunner have to do with any of this? It's only job is to stop after the first failure, so it' `pytest -x`.

ev-br commented:
Ah OK, this might make sense if pytest runs doctests one by one needs to override the reporting. Let's come back to this after the weekend.

ev-br commented on scpdt/impl.py:
```diff
@@ -375,10 +377,19 @@ class DebugDTRunner(DTRunner):
+                    r = super().run(test, compileflags=compileflags, out=out, clear_globs=False)
+                    if clear_globs:
+                        test.globs.clear()
+                    return r
```
To summarize our discussion offline: 
- either this context setting should happen in the `DTRunner` + inherited by `DebugDTRunner` --- then remove it from `testmod` and `testfile`;
- or this context manipulation belongs to the frontend and thus needs to be done in the plugin. 

Since the plugin seems to need to provide a specialized runner because of `out` being a list not a callable, it seems to make sense to move this logic in this runner.

Sheila-nk commented:
It would probably be a good idea to make `continue_on_failure` configurable through ini options.

ev-br commented on scpdt/impl.py:
```diff
@@ -375,7 +376,7 @@ class DebugDTRunner(DTRunner):
     """
     def run(self, test, compileflags=None, out=None, clear_globs=True):
-        r = super().run(test, compileflags, out, False)
+        r = super().run(test, compileflags=compileflags, out=out, clear_globs=False)
```
kwargs do make it read much better, thank you!

Now that I look at this, why does it drop `clear_globs` argument? This is pre-existing of course, can we pass it here nonetheless? At this point, probably best to just add an `TODO` comment and move on though.

ev-br commented on scpdt/plugin.py:
```diff
@@ -99,16 +102,16 @@ def collect(self):
         optionflags = doctest.get_optionflags(self)
-        runner = doctest._get_runner(
+
+        # We plugin scpdt's `DebugDTRunner`
```
```suggestion
        # We plug in scpdt's `DebugDTRunner`
```

Sheila-nk commented:
@ev-br  does `nameerror_after_exception` config attrbute serve the same purpose as the `continue_on_failure` variable? I can utilize that instead.

ev-br commented on scpdt/plugin.py:
```diff
@@ -99,16 +102,16 @@ def collect(self):
             verbose=False,
             optionflags=optionflags,
-            checker=_get_checker(),
-            continue_on_failure=doctest._get_continue_on_failure(self.config),
```
why are we dropping continue_on_failure?

ev-br replied:
https://github.com/ev-br/scpdt/pull/87#issuecomment-1677434321 please make sure that this test passes with the plugin

ev-br commented on scpdt/plugin.py:
```diff
@@ -159,6 +162,58 @@ def collect(self):
+
+
+def _get_runner(checker, verbose, optionflags):
+    import doctest
```
Nit: there is probably no reason to hide imports; move it to the module level.

Sheila-nk replied:
It's because of this error: 
```python
AttributeError: module '_pytest.doctest' has no attribute 'UnexpectedException'
```

ev-br commented on scpdt/plugin.py:
```diff
@@ -159,6 +162,58 @@ def collect(self):
+                        super().run(test, compileflags=compileflags, out=out, clear_globs=clear_globs)
+  
+        """
+        Almost verbatim copy of `_pytest.doctest.PytestDoctestRunner`.
```
Please detail the differences.

ev-br commented on scpdt/plugin.py:
```diff
@@ -159,6 +162,58 @@ def collect(self):
+            np_errstate: restores the numpy errstate and printoptions when done
+            user_context_manager: allows user to plug in their own context manager
+            matplotlib_make_nongui: temporarily make the matplotlib backend non-GUI
+            """
```
Please refer to `testmod` source for a discussion of the order of context managers.

Sheila-nk replied:
Can you please clarify what you mean by this? Do you want me to clean the comment or replace it with the comment in `testmod`?

ev-br replied:
Either way

ev-br commented on scpdt/plugin.py:
```diff
@@ -159,6 +162,58 @@ def collect(self):
+            If it's False, the failure or exception is raised immediately, potentially stopping further 
+            execution of the particular doctest
+            """
+            self.continue_on_failure = continue_on_failure
```
`config` argument is ignored, what is `self.config` below in `self.config.user_context_mgr`. This ties up to a question of where the global DTConfig lives.

ev-br reviewed (commented):
LGTM. 

Several comments in-line, none of them blocking. Please address those which are easy, and add TODO comments for the rest. 

The pytest code style is a little baroque with local imports and local classes, can move them to the module level in a clean-up. We should at some point also look into minimizing copy-paste from pytest, this is also for some other day.

There are likely several bugs lurking, these should be smoked out with dogfooding on scipy documentation, which should be the next step IMO.

ev-br commented:
> @ev-br does nameerror_after_exception config attrbute serve the same purpose as the continue_on_failure variable? I can utilize that instead.

Not sure actually. I don't know what `continue_on_failure` is; 
The job of `nameerror_after_exception` originally comes from the scipy issue linked in the commit message of 
 https://github.com/ev-br/scpdt/pull/35/commits/88d8ed710ab2cc35dfc0c8a99274fff7c6abbd1f .

Sheila-nk commented:
@ev-br 
I have made all the recommended changes except for:
1. the declaration of the `PytestDTRunner` class and import of `doctest` at the module level.
When I moved everything to the module level, it seems all references to `doctest` is from the `_pytest` module which brings up this error:
```python
AttributeError: module '_pytest.doctest' has no attribute 'UnexpectedException'
```
2. I have moved the creation of a test-case for the Name Error after exception to a new issue. I am yet to figure out what the test case looks like but manual testing by setting the `namerror_after_exception` config attribute in `conftest.py` is working as expected.

ev-br commented:
LGTM, thank you Sheila! 

Let's not worry about cleanups at this stage, I'll do a holistic review later anyway.
